### PR TITLE
Error for `typeof(T?)` when `T` is a reference type

### LINKED ISSUE #29894: Should nullable reference types be disallowed in `typeof`?
```
        [Fact]
        public void TypeOf_02()
        {
            CSharpCompilation c = CreateCompilation(new[] { @"
class List<T> { }
class C<T, TClass, TStruct>
    where TClass : class
    where TStruct : struct
{
    void M()
    {
        _ = typeof(C<int, object, int>?);
        _ = typeof(T?);
        _ = typeof(TClass?);
        _ = typeof(TStruct?);
        _ = typeof(List<T?>);
        _ = typeof(List<TClass?>);
        _ = typeof(List<TStruct?>);
    }
}
", NonNullTypesTrue, NonNullTypesAttributesDefinition });

            // PROTOTYPE(NullableReferenceTypes): should nullable reference types be disallowed in `typeof`?
            c.VerifyDiagnostics(
                // (10,20): error CS8627: A nullable type parameter must be known to be a value type or non-nullable reference type. Consider adding a 'class', 'struct', or type constraint.
                //         _ = typeof(T?);
                Diagnostic(ErrorCode.ERR_NullableUnconstrainedTypeParameter, "T?").WithLocation(10, 20),
                // (13,25): error CS8627: A nullable type parameter must be known to be a value type or non-nullable reference type. Consider adding a 'class', 'struct', or type constraint.
                //         _ = typeof(List<T?>);
                Diagnostic(ErrorCode.ERR_NullableUnconstrainedTypeParameter, "T?").WithLocation(13, 25)
                );
        }
```

### PR DESCRIPTION
**Pull request by gafter**
Fixes #29894

### CODE REVIEW

**Comment by gafter** [comment]
@dotnet/roslyn-compiler May I please have a second review for this nullable fix?

File: src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs [hunk 1]
```diff
@@ -89050,5 +89061,72 @@ class Derived<T> : Base<T> where T : class
...
+{
+    object o1 = typeof(int);
+    object o2 = typeof(string);
+    object o3 = typeof(int?);
+    object o4 = typeof(string?);    // 4, 5
+    object o5 = typeof(K<int?>);
+    object o6 = typeof(K<string?>); // 6
+    object o7 = typeof(K<int?>?);   // 7, 8
+    object o8 = typeof(K<string?>?);// 9, 10, 11
+}
```
Comment (jcouv) [inline src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs]: nit: would it also be worth capturing behavior for generic type parameters? (`T`, `TClass`, `TStruct`)

File: src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs [hunk 2]
```diff
@@ -33220,13 +33226,18 @@ void M()
...
                 //         _ = typeof(T?);
                 Diagnostic(ErrorCode.ERR_NullableUnconstrainedTypeParameter, "T?").WithLocation(10, 20),
+                // (11,13): error CS8639: The typeof operator cannot be used on a nullable reference type
+                //         _ = typeof(TClass?);
+                Diagnostic(ErrorCode.ERR_BadNullableTypeof, "typeof(TClass?)").WithLocation(11, 13),
                 // (13,25): error CS8627: A nullable type parameter must be known to be a value type or non-nullable reference type. Consider adding a 'class', 'struct', or type constraint.
                 //         _ = typeof(List<T?>);
-                Diagnostic(ErrorCode.ERR_NullableUnconstrainedTypeParameter, "T?").WithLocation(13, 25)
-                );
+                Diagnostic(ErrorCode.ERR_NullableUnconstrainedTypeParameter, "T?").WithLocation(13, 25));
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs]: >)); [](start = 102, length = 3)

Consider preserving original formatting.

File: src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs [hunk 3]
```diff
@@ -33220,13 +33226,18 @@ void M()
 
             // https://github.com/dotnet/roslyn/issues/29894: should nullable reference types be disallowed in `typeof`?
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs]: >// https://github.com/dotnet/roslyn/issues/29894: should nullable reference types be disallowed in `typeof`? [](start = 12, length = 108)

Can we remove this comment now?

**Pull request merged**